service/storage_service.cc

```diff
@@ -2089,6 +2224,78 @@ future<node_ops_cmd_response> storage_service::node_ops_cmd_handler(gms::inet_ad
+                    });
+                },
+                [&ss, ops_uuid ] { ss.node_ops_singal_abort(ops_uuid); });
+                ss._node_ops.emplace(ops_uuid, std::move(meta));
+            } else if (req.cmd == node_ops_cmd::replace_prepare_mark_alive) {
+                // Wait for local node has marked replacing node as alive
+                size_t nr_replacing_nodes = req.replace_nodes.size();
+                auto start_time = std::chrono::steady_clock::now();
+                auto timeout = std::chrono::seconds(120);
+                for (;;) {
+                    std::vector<gms::inet_address> replacing_nodes_live;
+                    for (auto& x: req.replace_nodes) {
+                        auto replacing_node = x.second;
+                        size_t nr_alive = ss.container().map_reduce0([replacing_node] (auto&& ss) -> size_t {
+                            return ss._gossiper.is_alive(replacing_node) ? 1 : 0;
+                        }, 0, std::plus<size_t>()).get0();
+                        slogger.debug("replace[{}]: Marked replacing_node={} as alive on {} out of {} shards", req.ops_uuid, replacing_node, nr_alive, smp::count);
+                        if (nr_alive == smp::count) {
+                            replacing_nodes_live.push_back(replacing_node);
+                        }
+                    }
+                    slogger.debug("replace[{}]: Waited for marking replacing node as up, replace_nodes={}, replacing_nodes_live={}",
+                            req.ops_uuid, req.replace_nodes, replacing_nodes_live);
+                    if (replacing_nodes_live.size() == nr_replacing_nodes) {
+                        break;
+                    }
+                    if (std::chrono::steady_clock::now() > timeout + start_time) {
+                        throw std::runtime_error(format("Failed to mark replacing node as alive in {} seconds, replace_nodes={}, replacing_nodes_live={}",
+                                timeout.count(), req.replace_nodes, replacing_nodes_live));
+                    }
+                    sleep_abortable(std::chrono::milliseconds(100), ss._abort_source).get();
+                }
+            } else if (req.cmd == node_ops_cmd::replace_prepare_pending_ranges) {
+                // Update the pending_ranges for the replacing node
+                slogger.debug("replace[{}]: Updated pending_ranges from coordinator={}", req.ops_uuid, coordinator);
+                ss.mutate_token_metadata([coordinator, &req, &ss] (mutable_token_metadata_ptr tmptr) mutable {
+                    return ss.update_pending_ranges(tmptr, format("replace {}", req.replace_nodes));
+                }).get();
+            } else if (req.cmd == node_ops_cmd::replace_heartbeat) {
+                slogger.debug("replace[{}]: Updated heartbeat from coordinator={}", req.ops_uuid, coordinator);
+                ss.node_ops_update_heartbeat(ops_uuid);
```

Comment:
The idea is to avoid introducing verbs more than necessary.

Reply:
I meant new node_ops_cmd (not RPC verb). We don't need separate cmds for every op, e.g. `removenode_done` and `replace_done`, since both just call "done" on a given op. So you can have a single command called  "done".

Reply:
For now, the handlers are same for removenode and replace, but in the future we might want to handle them differently, e.g., count how many ops were performed and when was finished.  It is better to future proof. The benefit of saving node_ops_cmd enum is tiny.

Reply:
The information about the kind of the operation is already known on the receiver, it is associated with a given uuid when the op is constructed. So you can still count per-op things even if we have a single command to terminate a given op.

It's not a big deal, it only seems like a duplication of effort to implement each of these separately.

Reply:
When a node restarts, it lost the uuid to ops mapping.